crates/matrix-sdk-ui/src/timeline/builder.rs

```diff
@@ -145,6 +145,14 @@ impl TimelineBuilder {
         self
     }
 
+    /// Use `VectorDiff`s as the new “input mechanism” for the `Timeline`.
+    ///
+    /// Read `TimelineSettings::vectordiffs_as_inputs` to learn more.
+    pub fn with_vectordiffs_as_inputs(mut self) -> Self {
+        self.settings.vectordiffs_as_inputs = true;
+        self
+    }
```

Comment:
I was thinking about this quite a lot too.

First off, this setting is temporary. It's gonna be removed as soon as we consider `RoomEventCacheUpdate::UpdateTimelineEvents` stable and we want to switch to it. The other variant, `AddTimelineEvents` will be removed at the same time.

Second, we have validated this approach during our team meeting :-p.

Third, while I share your concerns, putting the setting on the `EventCache` makes it so the `Timeline` can receive both inputs at the same time (`AddTimelineEvents` and `UpdateTimelineEvents`), which can create a lot of confusion and inconsistent state.

Fourth, this setting is used to decide if the `Timeline` has to apply its own deduplication mechanism or not:
- with `AddTimelineEvents`: yes, it has to,
- with `UpdateTimelineEvents: no, because it's expected to be handle by the `VectorDiff`.
Putting the settings on the `EventCache` makes it way more difficult for the `Timeline` to decide. Actually, it requires a setting in the `Timeline` anyway.

Thoughts?

Reply:
> Third, while I share your concerns, putting the setting on the EventCache makes it so the Timeline can receive both inputs at the same time (AddTimelineEvents and UpdateTimelineEvents), which can create a lot of confusion and inconsistent state.

No: the idea is that, the EventCache would emit `AddTimelineEvents` if `!feature_flag`, and emit `UpdateTimelineEvents` if `feature_flag`, not both at the same time.

Re: fourth, you're right the timeline would need to get this information; it's relatively trivial to add a getter on the feature flag in the event cache, and get the feature flag value from the `TimelineBuilder`.

(Re: second, sometimes people change their mind when they realize they were wrong, and I definitely was when I validated it :innocent:)

Overall, I would even call for *not* having a feature flag at all, and remove the previous code entirely, but I understand that it's a big change, and it would be nice to test in isolation (especially before the EOY holidays). So my preference would be to have a feature flag on the EventCache, for consistency (it would also simplify code at the FFI layer, I suppose), but we can probably live with the feature flag on the Timeline, if you strongly prefer it.